src/Language/Marlowe/Runtime/Types/ContractsJSON.hs

```diff
@@ -32,20 +48,98 @@ instance FromJSON ContractInList where
+
+
+getContracts :: String -> [ContractInList] -> IO (Either String ContractList)
+getContracts endpoint lOldChain = do
+  eresult <- runGetContracts (endpoint, fromList lOldChain) $ getContracts' (empty, Start)
+  now <- getCurrentTime
+  return $ (Right . ContractList now . toList . fst) =<< eresult
+
+
+-- This code has two criteria when trying to get the latest Marlowe contract
+-- data to add to what we already have cached.
+--
+-- 1. We need to keep asking the Marlowe Runtime for pages of contract data
+--    until the Next-Range header indicates there are no more.
+-- 2. Regardless of the Runtime paging above, we need to only download
+--    contracts from the tip backwards until we locate the point where our
+--    prior contract list has them already. This will handle both new contracts
+--    since the last time and also any rollbacks that have occurred, because we
+--    trim off the cached chain prior to the point where it matches new data.
+--
+-- Point 1 is what the Range (Start | Next ... | Done) data structure is for
+-- Point 2 is what the Language.Marlowe.Runtime.AdjustTip.BuildChain
+-- (Building | BuildDone) data structure is for
+
+getContracts' :: (Seq ContractInList, Range) -> GetContracts (Seq ContractInList, Range)
+
+getContracts' t@(_acc, Done) = return t
+
+getContracts' (acc, range) = do
+  (nextContracts, nextRange) <- contractsRESTCall range
+  oldChain <- asks snd
+  case processTip oldChain nextContracts of
+    -- Note setting Range to Done here because we have enough info and can stop
+    BuildDone endOfChain -> getContracts' (acc >< endOfChain, Done)
+    Building newTip -> getContracts' (acc >< newTip, nextRange)
+
+
+setRangeHeader :: Range -> Request -> Request
+setRangeHeader (Next bs) = setRequestHeader "Range" [bs]
+setRangeHeader _ = id
+
+parseRangeHeader :: [ByteString] -> Range
+parseRangeHeader [bs] = Next bs
+parseRangeHeader _ = Done
 
-getContracts :: String -> IO (Either String ContractList)
-getContracts endpoint = do
+contractsRESTCall :: Range -> GetContracts (Seq ContractInList, Range)
+contractsRESTCall range = do
+  endpoint <- asks fst
   initialRequest <- parseRequest $ endpoint <> "contracts"
-  let request = setRequestMethod "GET" $ setRequestHeader "Accept" ["application/json"] initialRequest
-  response <- httpLBS request
-  return $ eitherDecode (getResponseBody response)
+  let request = foldl' (flip id) initialRequest
+        [ setRequestMethod "GET"
+        , setRequestHeader "Accept" ["application/json"]
+        , setRangeHeader range
+        ]
+  eresponse <- liftIO . try $ httpLBS request
+  case eresponse of
+    Left exception -> throwError . show $ (exception :: HttpException)
+    Right response -> case eitherDecode (getResponseBody response) of
+      Left err -> throwError err
+      Right (InternalContractList contracts) -> do
+        let nextRange = parseRangeHeader . getResponseHeader "Next-Range" $ response
+        return (fromList contracts, nextRange)
```

Comment:
About the `ReaderT` env, yes, will change to a record

I think there isn't really an `EitherT`, well there is in a library we're not using `transformers-either`.
`ExceptT` we are using uses `Either e` as its monad but the type we get back from `httpLBS` isn't the same exactly as the rest of our computation.
BUT I could probably simplify this case statement with `liftEither` and `modifyError` from `Control.Monad.Except`. The `Left` case is about transforming the error type. After it will probably be a nicer-looking bind situation.

Reply:
Oh wait, it's becoming an `Either` because of the `try` I'm using on the `httpLBS` IO action. Hm. I may be able to use something different from `Control.Exception`

Reply:
Maybe a new ticket?